mbf_abstract_nav/src/move_base_action.cpp

```diff
@@ -406,18 +393,22 @@ void MoveBaseAction::actionExePathDone(
           }
           else
           {
-            ROS_WARN_STREAM_NAMED("move_base", "Abort the execution of the controller: " << result.message);
+            ROS_WARN_STREAM_NAMED("move_base", "Abort the execution of the controller: " << exe_path_result.message);
             goal_handle_.setAborted(move_base_result, state.getText());
           }
           break;
       }
       break;
 
     case actionlib::SimpleClientGoalState::PREEMPTED:
-      // action was preempted successfully!
-      ROS_DEBUG_STREAM_NAMED("move_base", "The action \""
-          << "exe_path" << "\" was preempted successfully!");
-      // TODO
+      ROS_INFO_STREAM_NAMED("move_base", "The action \"" << "exe_path" << "\" has been preempted!");
+      if (action_state_ == CANCELED)
```

Comment:
move base has been preempted. otherwise we have preempted internally, (what I don't think can happen) so we don't end the action

Reply:
i think checking for CANCELED is contra productive. 

We may end up at PREEMPTED by either calling SimpleActionClient::cancelGoal() or by someone else kicking us out of the current execution (see http://wiki.ros.org/actionlib/DetailedDescription)

In this class, we only call SimpleActionClient::cancelGoal() of action_client_get_path_ and action_client_recovery_ clients from MoveBaseAction::cancel(). So there is no "internal" preemption, and by removing this check we may react to another client stopping our goal. 

For action_client_exe_path_ its a bit more tricky, since we call SimpleActionClient::cancelGoal() on oscillation (so there is an internal preemption). What we might do is to reuse the recovery_trigger_ member to check inside the MoveBaseAction::actionExePathDone() if we want to react to a PREEMPTED flag or not:

```
void MoveBaseAction::actionExePathFeedback(
    const mbf_msgs::ExePathFeedbackConstPtr &feedback)
{
  // ...
  if (!oscillation_timeout_.isZero())
  {
    if (mbf_utility::distance(robot_pose_, last_oscillation_pose_) >= oscillation_distance_)
    {
    // ...
    else if (last_oscillation_reset_ + oscillation_timeout_ < ros::Time::now())
    {
      // ...
      // use this to indicate that we don't react to failures
      recovery_trigger_ = OSCILLATING;
      
      action_client_exe_path_.cancelGoal();

      if (!attemptRecovery())
      {
       // ...
      }
    }
  }
}

```
Then 
```
void MoveBaseAction::actionExePathDone(...){
// ...
switch(state.state_){
case PREEMPTED:
if(recovery_trigger_ != OSCILLATION) 
   goal_handle_.setCanceled(...);
}
}
```
And finally
```
void MoveBaseAction::actionRecoveryDone(...){
...
recovery_trigger_ = None;
}
```

Reply:
> i think checking for CANCELED is contra productive.
> 
> We may end up at PREEMPTED by either calling SimpleActionClient::cancelGoal() or by someone else kicking us out of the current execution (see http://wiki.ros.org/actionlib/DetailedDescription)
Don't think so, cause we only set action_state_ = CANCELED on cancel() method, and it only gets called we we cancel current move_base goal, not when preempting with a new one.

Reply:
my point is that if someone else preempts the current exec (by calling for example directly the exe_path action, instead of the move_base action), then the action_client_exe_path_ would signal PREEMPTED but the action_state_ would be not set to CANCELED. Then the client of the move_base action would hang

Reply:
I tried to do so using the ugly script below and it works, both master and this branch. Please give it a try and ideally modify it to trigger the cases you are worried about.

```
#!/usr/bin/env python

import time
import random

import rospy
import actionlib

from geometry_msgs.msg import PoseStamped

from mag_common_py_libs.geometry import quaternion_from_yaw, pose2str

from mbf_msgs.msg import GetPathAction, GetPathGoal, GetPathResult
from mbf_msgs.msg import ExePathAction, ExePathGoal, ExePathResult
from mbf_msgs.msg import MoveBaseAction, MoveBaseGoal, MoveBaseResult

__author__ = 'santos'

""" 
Simplest possible MBF executive: sends goals provided through RViz on 2D Nav Goal tool to
move_base_flex/get_path action and sends the resulting path to move_base_flex/exe_path action
No fancy recovery or retry strategies.
"""


def create_geo_pose_stamped(x, y, yaw, frame_id='map'):
    pose = PoseStamped()
    pose.pose.position.x = x
    pose.pose.position.y = y
    pose.pose.orientation = quaternion_from_yaw(yaw)
    pose.header.frame_id = frame_id
    return pose

def nav_goal_cb(msg):
    global target_pose
    target_pose = msg
    goal = GetPathGoal(use_start_pose=False,
                       start_pose=msg,
                       target_pose=msg)
    get_path_ac.send_goal(goal, done_cb=plan_done_cb)
    rospy.loginfo("Relaying move_base_simple/goal %s from RViz to MBF's planner", pose2str(msg))
    time.sleep(1)
    goal = MoveBaseGoal(target_pose=msg,
                        controller='TEBPlanner')
    move_base_ac.send_goal(goal, done_cb=move_done_cb)

def plan_done_cb(status, result):
    if result.outcome == GetPathResult.SUCCESS:
        rospy.loginfo("Get path action succeeded")
        time.sleep(5)

        rospy.loginfo("Now Exe path !!!!!!!!!")
        goal = ExePathGoal(path=result.path,
                           controller='DWAPlanner')
        exe_path_ac.send_goal(goal, done_cb=follow_done_cb)
        target_pose = None
    else:
        rospy.logerr("Get path action failed with error code [%d]: %s", result.outcome, result.message)

def follow_done_cb(status, result):
    if result.outcome == GetPathResult.SUCCESS:
        rospy.loginfo("Follow path action succeeded")
    else:
        rospy.logerr("Follow path action failed with error code [%d]: %s", result.outcome, result.message)

def move_done_cb(status, result):
    if result.outcome == MoveBaseResult.SUCCESS:
        rospy.loginfo("Follow path action succeeded")
    else:
        rospy.logerr("Follow path action failed with error code [%d]: %s", result.outcome, result.message)

if __name__ == '__main__':
    rospy.init_node("simple_mbf_executive")

    target_pose = None

    get_path_ac = actionlib.SimpleActionClient("/move_base_flex/get_path", GetPathAction)
    get_path_ac.wait_for_server(rospy.Duration(5))
    exe_path_ac = actionlib.SimpleActionClient("/move_base_flex/exe_path", ExePathAction)
    exe_path_ac.wait_for_server(rospy.Duration(5))
    move_base_ac = actionlib.SimpleActionClient("/move_base_flex/move_base", MoveBaseAction)
    move_base_ac.wait_for_server(rospy.Duration(5))

    rospy.Subscriber('/move_base_simple/goal', PoseStamped, nav_goal_cb)

    rospy.spin()
```